fix(core): update `ApplicationRef.isStable` to account for rendering pending tasks

alan-agius4 opened the pull request:
**fix(core): update `ApplicationRef.isStable` to account for rendering pending tasks**
    
This commit updates the `ApplicationRef.isStable` API to account for pending rendering task. This is needed as once a pending rendering task is done, new macrotask and microtask could be created which previously caused these not to be intercepted and thus ignored when doing SSR.


**refactor(http): replace zone.js macrotask creation with `InitialRenderPendingTasks`**

This commits refactors the HTTP client to use `InitialRenderPendingTasks` instead of Zone.js macrotask. This is another approach to https://github.com/angular/angular/pull/50406 which was revert due to a failure in G3.

alfaproject commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -165,16 +165,14 @@ export function withHttpTransferCache(): Provider[] {
         };
       },
-      deps: [ApplicationRef, CACHE_STATE, InitialRenderPendingTasks]
+      deps: [ApplicationRef, CACHE_STATE]
```
This shouldn't be needed anymore because there's no arguments in the factory function

AndrewKushnir reviewed (dismissed):
LGTM 👍 

@alan-agius4 my original thinking was that we wanted to do the inverse though: use the `InitialRenderPendingTasks` as the main instance that controls the state and make `ApplicationRef.isStable` contribute to it. However, in this case the `InitialRenderPendingTasks` needs to be an observable too. Let's ship the fix as is and come back to this discussion later (as a part of zoneless support for SSR).

alan-agius4 committed:
fix(core): update `ApplicationRef.isStable` to account for rendering pending tasks

This commit updates the `ApplicationRef.isStable` API to account for
pending rendering task. This is needed as once a pending rendering task
is done, new macrotask and microtask could be created which previously caused these not
to be intercepted and thus ignored when doing SSR.

alan-agius4 committed:
refactor(http): replace zone.js macrotask creation with `InitialRenderPendingTasks`

This commits refactors the HTTP client to use `InitialRenderPendingTasks` instead of Zone.js macrotask. This is another approach to https://github.com/angular/angular/pull/50406 which was revert due to a failure in G3.

alan-agius4 commented:
- [TGP](http://fusion2/presubmit/535202133/OCL:535202133:BASE:535626215:1685115396822:b517c4dc/targets)
- [Guitar](http://fusion2/95b42796-e2d7-3e99-8a43-4fe5a7616a4e) run for previous failure in G3 caused by https://github.com/angular/angular/pull/50406

alan-agius4 committed:
fixup! fix(core): update `ApplicationRef.isStable` to account for rendering pending tasks

AndrewKushnir reviewed (approved):
@alan-agius4 this change looks awesome! 👍 Thank you.

alan-agius4 committed:
fixup! fix(core): update `ApplicationRef.isStable` to account for rendering pending tasks

alan-agius4 commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -52,7 +52,7 @@
     "uncompressed": {
       "runtime": 918,
-      "main": 87081,
+      "main": 91485,
```
The difference introduced with this change is about ~1.5K.

alan-agius4 commented:
**Caretaker note:** G3 failures are pre-existing.

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

AndrewKushnir commented:
@alan-agius4 I've LGTM'ed for `size-tracking`, is it ready for merge?

alan-agius4 commented:
> 

Yes @AndrewKushnir

dylhunn commented:
This PR was merged into the repository by commit 0c80349cf649256ba96ee6b60073fce24f54677c.

alan-agius4 commented:
@matthew-whitfield-uk, please subscribe to https://github.com/angular/angular/issues/50562
